src/VirtoCommerce.CatalogExportImportModule.Data/Services/CsvPagedPhysicalProductDataImporter.cs

```diff
@@ -315,5 +321,35 @@ private void SetCatalogId(CatalogProduct[] productsToSave, string catalogId)
                 product.CatalogId = catalogId;
             }
         }
+
+        private void SetMainProductIdFromTheOuterIfMainValueIsBad(ImportRecord<CsvPhysicalProduct>[] records, CatalogProduct[] existedMainProducts)
+        {
+            foreach (var record in records.Where(x => !string.IsNullOrEmpty(x.Record.MainProductId) || !string.IsNullOrEmpty(x.Record.MainProductOuterId)))
+            {
+                // Try to find by MainProductId
+                var mainProductId = record.Record?.MainProductId;
+                var existedMainProduct =
+                    existedMainProducts.FirstOrDefault(x => x.Id.EqualsInvariant(mainProductId));
+
+                if (existedMainProduct is null)
+                {
+                    // If fails, then try to find by MainProductOuterId
+                    var mainProductOuterId = record.Record?.MainProductOuterId;
+
+                    if (!string.IsNullOrEmpty(mainProductOuterId))
+                    {
+                        existedMainProduct = existedMainProducts.FirstOrDefault(x => x.OuterId.EqualsInvariant(mainProductOuterId));
+                    }
+
+                    // If find by outer id, then replace MainProductId value
+                    record.Record.MainProductId = existedMainProduct?.Id ?? record.Record.MainProductId;
```

Comment:
What if MainProductId is empty and MainProductOuterId is invalid? Shouldn't it be `?? MainProductId ?? MainProductOuterId` and, probably with check IsNullOrEmpty and not simple `??`